src/test/scala/org/ergoplatform/mining/ErgoMinerSpec.scala

```diff
@@ -60,7 +71,135 @@ class ErgoMinerSpec extends TestKit(ActorSystem()) with FlatSpecLike with ErgoTe
       blocksGenerated should be > 1
       blocksGenerated should be >= height
     }
+
+    system.terminate()
+  }
+
+  it should "filter out double spend txs" in {
+    val tx = validErgoTransactionGen.sample.get._2
+    ErgoMiner.fixTxsConflicts(Seq(tx, tx, tx)) should have length 1
+
+    val inputs = validErgoTransactionGenTemplate(0, -1, 100, 100).sample.get._1
+    val (l, r) = inputs.splitAt(50)
+    val tx_1 = validTransactionGen(l).sample.get
+    val tx_2 = validTransactionGen(r :+ l.last).sample.get
+
+    ErgoMiner.fixTxsConflicts(Seq(tx_1, tx_2, tx)) should contain theSameElementsAs Seq(tx_1, tx)
+    ErgoMiner.fixTxsConflicts(Seq(tx_2, tx_1, tx)) should contain theSameElementsAs Seq(tx_2, tx)
+  }
+
+  it should "create own coinbase transaction, if there is already a transaction, that spends emission box" in new TestKit(ActorSystem()){
+    val tmpDir = createTempDir
+
+    type msgType = SemanticallySuccessfulModifier[_]
+    val newBlock = classOf[msgType]
+    val testProbe = new TestProbe(system)
+    system.eventStream.subscribe(testProbe.ref, newBlock)
+    val newBlockDuration = 30 seconds
+
+    val defaultSettings: ErgoSettings = ErgoSettings.read(None).copy(directory = tmpDir.getAbsolutePath)
+
+    val nodeSettings = defaultSettings.nodeSettings.copy(mining = true,
+      stateType = StateType.Utxo,
+      miningDelay = defaultAwaitDuration,
+      offlineGeneration = true,
+      verifyTransactions = true)
+    val chainSettings = defaultSettings.chainSettings.copy(blockInterval = 2.seconds)
+    val ergoSettings = defaultSettings.copy(nodeSettings = nodeSettings, chainSettings = chainSettings)
+    val nodeViewHolderRef: ActorRef = ErgoNodeViewRef(ergoSettings, timeProvider, emission)
+    val readersHolderRef: ActorRef = ErgoReadersHolderRef(nodeViewHolderRef)
+
+    val minerRef: ActorRef = ErgoMinerRef(ergoSettings, nodeViewHolderRef, readersHolderRef, timeProvider, emission, TrueLeaf)
+    expectNoMessage(1 second)
+    val r: Readers = Await.result((readersHolderRef ? GetReaders).mapTo[Readers], 10 seconds)
+
+    val emissionBox = r.s.asInstanceOf[UtxoStateReader].emissionBoxOpt.get
+
+    val prop1 = DLogProverInput(BigIntegers.fromUnsignedByteArray("test1".getBytes())).publicImage
+
+    val input = Input(emissionBox.id, ProverResult(Array.emptyByteArray, ContextExtension.empty))
+    val outputs = IndexedSeq(new ErgoBoxCandidate(1000000L, prop1))
+    val tx = new ErgoTransaction(IndexedSeq(input), outputs)
+
+    nodeViewHolderRef ! LocallyGeneratedTransaction[ErgoTransaction](tx)
```

Comment:
Please add a check that it was added to mempool

Reply:
@catena2w It won't go to mempool, it won't pass validation.